man/jar-awk.1
.TH JAR-AWK 1 "May 4, 2014" 
.SH NAME
.PP
jar-awk - read multiline records from a file or standard input and
process them with Ruby
.SH SYNOPSIS
.PP
jar-awk [-t] -l REGEX [-F REGEX] [-b RUBY_SCRIPT] [-e RUBY_SCRIPT] (-f
RUBY_PATH | RUBY_SCRIPT) [JAR_INPUT] ...
.SH DESCRIPTION
.PP
\f[C]jar-awk\f[] processes multiline records in a file or input stream.
The records are separated by lines which match the REGEX expression
specified by the \f[C]-l\f[] (\f[C]--line-delimiter\f[]) flag.
.PP
\f[C]jar-awk\f[] processes two formats: \f[C]cookie\ jars\f[] and
\f[C]record\ jars\f[].
These terms from the section on data file formats in the
\f[C]The\ Art\ of\ Unix\ Programming\f[].
.PP
In the cookie jar format, the records are just strings of text that can
contain new lines.
\f[C]jar-awk\f[] sets the variable \f[C]$_\f[] to each record in turn.
These cookie jars have nothing to do with browser cookies.
.PP
If the \f[C]-F\f[] (\f[C]--field-delimiter\f[]) flag is used, the input
is assumed to be in record jar format.
Each line of the record contains a key-value pair.
The argument of the \f[C]-F\f[] flag is a REGEX which is used to split
the line in two.
\f[C]jar-awk\f[] sets \f[C]$_\f[] to a Ruby hash containing the
key-value pairs.
.PP
The Ruby script can be provided on the command line to do something with
\f[C]$_\f[] for each record.
Optionally, the \f[C]-f\f[] flag can be used to read the Ruby script
from a file.
.PP
Some cookie jar files and record jar files may put interesting
information on the delimiter line.
The match data for the delimiter line is put in the variable
\f[C]$md\f[].
If the regular expression contained groups, these are available in
\f[C]$md[1]\f[], \f[C]$md[2]\f[], and so on.
The entire matched expression is in \f[C]$md[0]\f[], and what came
before and after it is in \f[C]$md.pre_match\f[] and
\f[C]$md.post_match\f[].
.PP
Because \f[C]jar-awk\f[] is implemented with \f[C]eval\f[] instead of
\f[C]ruby\ -e\f[], \f[C]BEGIN\f[] and \f[C]END\f[] blocks are not
supported.
To specify blocks of Ruby code which execute at the begin and end of
execution, use the \f[C]-B\f[] (\f[C]--BEGIN\f[]) and \f[C]-E\f[]
(\f[C]--END\f[]) flags.
.SH OPTIONS
.PP
-B RUBY_SCRIPT, --BEGIN=RUBY_SCRIPT : execute Ruby script provided as
argument before processing records
.PP
-E RUBY_SCRIPT, --END=RUBY_SCRIPT : execute Ruby script provided as
argument after processing records
.PP
-f PATH, --file=PATH : execute the Ruby script in the file at PATH.
.PP
-F REGEX, --field-delimiter=REGEX : the regular expression
.PP
-l REGEX, --line-delimiter=REGEX : a regular expression for identifying
lines between records.
.PP
-s, --strict : exit with error message and nonzero status when a record
line cannot be parsed as a key and value; this only has an effect with
the -F flag.
.PP
-S, --silent : do not emit an error message when the -F flag is in
effect and a record line cannot be parsed as a key and value.
.PP
-t, --trim : Trim whitespace off the edges of records.
Note that when the -F flag is use, whitespace is trimmed off the edges
of keys and values by defualt.
.PP
-T, --no-trim : When the -F flag is specified, do not trim whitespace
off the edges of keys and values.
.PP
-Z, --skip-record-zero : If there are lines before the first line which
matches the -l REGEX, skip them.
By default they are treated as a record
.SH EXAMPLES
.PP
On Mac OS X, list the sections of the \f[C]system_profiler\f[] output:
.IP
.nf
\f[C]
system_profiler\ -detailLevel\ mini\ |\ grep\ -v\ \[aq]^\\s*$\[aq]\ |\ jar-awk\ -l\ \[aq]^(\\S+):\[aq]\ \[aq]puts\ $md[1]\[aq]
\f[]
.fi
.PP
On Mac OS X, extract the \f[C]Software\f[] section of the
\f[C]system_profiler\f[] output:
.IP
.nf
\f[C]
system_profiler\ -detailLevel\ mini\ |\ grep\ -v\ \[aq]^\\s*$\[aq]\ |\ jar-awk\ -l\ \[aq]^(\\S+):\[aq]\ \[aq]puts\ $_\ if\ $md[1]\ ==\ "Software"\[aq]
\f[]
.fi
.PP
One Mac OS X, extract the \f[C]System\ Version\f[] from the
\f[C]Software\f[] section of the \f[C]system_profiler\f[] output:
.IP
.nf
\f[C]
system_profiler\ -detailLevel\ mini\ |\ grep\ -v\ \[aq]^\\s*$\[aq]\ |\ jar-awk\ -t\ -F\ \[aq]:\[aq]\ -l\ \[aq]^(\\S+):\[aq]\ \[aq]puts\ $_["System\ Version"]\ if\ $md[1]\ ==\ "Software"\[aq]
\f[]
.fi
.SH SEE ALSO
.PP
\f[C]awk\f[] (1)
.PP
\f[C]csplit\f[] (1)
.PP
http://www.catb.org/esr/writings/taoup/html/ch05s02.html
.SH AUTHORS
Clark Grubb.


src/jar-awk.rb
#!/usr/bin/env ruby

require 'getoptlong'

def usage
    $stderr.puts "USAGE: jar-awk [-t] -l REGEX [-F REGEX] [-B SCRIPT] [-E SCRIPT] (-f SCRIPT_FILE | SCRIPT) [RECORD_FILE]"
    $stderr.puts
    $stderr.puts "  -t (--trim): trim whitespace"
    $stderr.puts "  -l (--line-delimiter): regex identifying lines between records"
    $stderr.puts "  -B (--BEGIN): script to execute before processing records"
    $stderr.puts "  -E (--END): script to execute after processing records"
    $stderr.puts "  -f (--file): execute script in file for each record instead of script on cmd line"
    $stderr.puts
    $stderr.puts "  $_: variable containing record"
    $stderr.puts "  $md: variable containing match data for -l REGEX"
    exit 1
end

opts = GetoptLong.new(
                      ['--BEGIN', '-B',
                       GetoptLong::REQUIRED_ARGUMENT],
                      ['--END', '-E',
                       GetoptLong::REQUIRED_ARGUMENT],
                      ['--file', '-f',
                       GetoptLong::REQUIRED_ARGUMENT],
                      ['--line-delimiter', '-l',
                       GetoptLong::REQUIRED_ARGUMENT],
                      ['--field-delimiter', '-F',
                       GetoptLong::OPTIONAL_ARGUMENT],
                      ['--strict', '-s',
                       GetoptLong::NO_ARGUMENT],
                      ['--silent', '-S',
                       GetoptLong::NO_ARGUMENT],
                      ['--trim', '-t',
                       GetoptLong::NO_ARGUMENT],
                      ['--no-trim', '-T',
                       GetoptLong::NO_ARGUMENT],
                      ['--skip-record-zero', '-Z',
                       GetoptLong::NO_ARGUMENT],
                      ['--help',
                       GetoptLong::NO_ARGUMENT]
                      )


script = nil
line_delimiter = nil
begin_script = nil
end_script = nil
use_zero = true

$strict = false
$silent = false
$field_delimiter = nil
$trim = nil

opts.each do |opt, arg|
  case opt
  when '--BEGIN'
    begin_script = arg
  when '--END'
    end_script = arg
  when '--file'
    script = File.open(arg).read
  when '--field-delimiter'
    $field_delimiter = /#{arg}/
  when '--line-delimiter'
    line_delimiter = /#{arg}/
  when '--strict'
    $strict = true
  when '--silent'
    $silent = true
  when '--no-trim'
    $trim = false
  when '--trim'
    $trim = true
  when '--skip-record-zero'
    use_zero = false
  when '--help'
    usage
  end
end

if $trim.nil?
  $trim = $field_delimiter ? true : false
end

if not script
  if ARGV.size > 0
    script = ARGV.shift
  else
    $stderr.puts "ERROR: no SCRIPT or SCRIPT_FILE specified"
    usage
  end
end

if not line_delimiter
  $stderr.puts "ERROR: no --line-delimiter or -l specified"
  usage
end

def get_binding
  binding
end

$binding = get_binding
$processor = eval("lambda { |line, md| $_ = line; $md = md; #{script} }", $binding)

if ARGV.size > 0
  input_stream = File.open(ARGV[0])
else
  input_stream = $stdin
end

class EmptyMatchData
  def pre_match; ''; end
  def post_match; ''; end
  def [](i); nil; end
  def length; 0; end
  def size; 0; end
  def names; []; end
end

class Record
  def initialize(match_data)
    @match_data = match_data
    if $field_delimiter
      @record = {}
    else
      @record = ''
    end
  end

  def add_line(line)
    if $field_delimiter
      key, value = line.split($field_delimiter, 2)
      if value.nil? and not $silent and not $strict
        $stderr.puts "WARNING: record line does not have field delimiter: #{line}"
      end
      if value.nil? and $strict
        $stderr.puts "ERROR: line does not have field delimiter: #{line}"
        exit 1
      end
      if $trim
        key.strip!
        value.strip! if value
      end
      @record[key] = value
    else
      @record += line
    end
  end

  def process
    if $trim and not $field_delimiter
      @record.strip!
    end
    $processor.call(@record, @match_data)
  end

end

if begin_script
  eval(begin_script, $binding)
end

record_num = 0
record = Record.new(EmptyMatchData.new)

input_stream.each_with_index do |line, line_num|
  md = line_delimiter.match(line)
  if md
    if (use_zero and record_num == 0 and line_num > 0) or record_num > 0
      record.process
    end
    record_num += 1
    record = Record.new(md)
  else
    record.add_line(line)
  end
end

if (use_zero and record_num == 0 and line_num > 0) or record_num > 0
  record.process
end

if end_script
  eval(end_script, $binding)
end
